build.js

```diff
@@ -0,0 +1,423 @@
+const path = require('path')
+const Hyperschema = require('hyperschema')
+
+const DIR = path.join(__dirname, 'encoding')
+const SPEC = path.join(DIR, 'spec/autobase')
+
+const schema = Hyperschema.from(SPEC, { versioned: true })
+const autobase = schema.namespace('autobase')
+
+autobase.require(path.join(DIR, 'legacy.js'))
+
+autobase.register({
+  name: 'checkout',
+  compact: true,
+  fields: [
+    {
+      name: 'key',
+      type: 'fixed32',
+      required: true
+    },
+    {
+      name: 'length',
+      type: 'uint',
+      required: true
+    }
+  ]
+})
+
+autobase.register({
+  name: 'clock',
+  array: true,
+  type: '@autobase/checkout'
+})
+
+autobase.register({
+  name: 'index-checkpoint',
+  compact: true,
+  fields: [
+    {
+      name: 'signature',
+      type: 'fixed64',
+      required: true
+    },
+    {
+      name: 'length',
+      type: 'uint',
+      required: true
+    }
+  ]
+})
+
+autobase.register({
+  name: 'wakeup',
+  external: 'Wakeup'
+})
+
+autobase.register({
+  name: 'boot-record-v0',
+  external: 'BootRecordV0'
+})
+
+autobase.register({
+  name: 'boot-record-raw',
+  fields: [
+    {
+      name: 'key',
+      type: 'fixed32',
+      required: true
+    },
+    {
+      name: 'systemLength',
+      type: 'uint',
+      required: true
+    },
+    {
+      name: 'indexersUpdated',
+      type: 'bool',
+      required: true
+    },
+    {
+      name: 'fastForwarding',
+      type: 'bool',
+      required: true
+    },
+    {
+      name: 'recoveries',
+      type: 'uint',
+      required: false
+    }
+  ]
+})
+
+autobase.register({
+  name: 'boot-record',
+  versions: [
+    {
+      version: 0,
+      type: '@autobase/boot-record-v0'
+    },
+    {
+      version: 3,
+      type: '@autobase/boot-record-raw'
+    }
+  ]
+})
+
+autobase.register({
+  name: 'checkpointer',
+  compact: true,
+  fields: [
+    {
+      name: 'checkpointer',
+      type: 'uint',
+      required: true
+    },
+    {
+      name: 'checkpoint',
+      type: '@autobase/index-checkpoint',
+      required: false
+    }
+  ]
+})
+
+autobase.register({
+  name: 'checkpoint',
+  compact: false,
+  fields: [
+    {
+      name: 'system',
+      type: '@autobase/checkpointer',
+      required: false
+    },
+    {
+      name: 'encryption',
+      type: '@autobase/checkpointer',
+      required: false
+    },
+    {
+      name: 'user',
+      type: '@autobase/checkpointer',
+      array: true,
+      required: false
+    }
+  ]
+})
+
+autobase.register({
+  name: 'digest',
+  compact: false,
+  fields: [
+    {
+      name: 'pointer',
+      type: 'uint',
+      required: true
```

Comment:
lets declare this optional, saves us a byte when key is set (cause would be 0) and no downside